crml/cennzx-spot/rpc/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+use jsonrpc_core::{Error as RpcError, ErrorCode, Result};
+use jsonrpc_derive::rpc;
+use sp_api::ProvideRuntimeApi;
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT, traits::SimpleArithmetic};
+
+pub use self::gen_client::Client as CennzxSpotClient;
+pub use crml_cennzx_spot_rpc_runtime_api::{
+	self as runtime_api, CennzxSpotApi as CennzxSpotRuntimeApi, CennzxSpotResult,
+};
+
+/// Contracts RPC methods.
+#[rpc]
+pub trait CennzxSpotApi<AssetId, Balance> {
+	#[rpc(name = "cennzx_buyPrice")]
+	// TODO: prefer to return Result<Balance>, however Serde JSON library only allows u64.
+	//  - change to Result<Balance> once https://github.com/serde-rs/serde/pull/1679 is merged
+	fn buy_price(&self, asset_to_buy: AssetId, amount_to_buy: Balance, asset_to_pay: AssetId) -> Result<u64>;
+
+	#[rpc(name = "cennzx_salePrice")]
+	// TODO: change to Result<Balance> once https://github.com/serde-rs/serde/pull/1679 is merged
+	fn sell_price(&self, asset_to_sell: AssetId, amount_to_buy: Balance, asset_to_payout: AssetId) -> Result<u64>;
+}
+
+/// An implementation of CENNZX Spot Exchange specific RPC methods.
+pub struct CennzxSpot<C, T> {
+	client: Arc<C>,
+	_marker: std::marker::PhantomData<T>,
+}
+
+impl<C, T> CennzxSpot<C, T> {
+	/// Create new `CennzxSpot` with the given reference to the client.
+	pub fn new(client: Arc<C>) -> Self {
+		CennzxSpot {
+			client,
+			_marker: Default::default(),
+		}
+	}
+}
+
+/// Error type of this RPC api.
+pub enum Error {
+	/// The call to runtime failed.
+	Runtime,
+	CannotExchange,
+	PriceOverflow,
+}
+
+impl From<Error> for i64 {
+	fn from(e: Error) -> i64 {
+		match e {
+			Error::Runtime => 1,
+			Error::CannotExchange => 2,
+			Error::PriceOverflow => 3,
+		}
+	}
+}
+
+impl<C, Block, AssetId, Balance> CennzxSpotApi<AssetId, Balance> for CennzxSpot<C, Block>
+where
+	Block: BlockT,
+	C: Send + Sync + 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+	C::Api: CennzxSpotRuntimeApi<Block, AssetId, Balance>,
+	AssetId: Codec,
+	Balance: Codec + SimpleArithmetic,
+{
+	fn buy_price(&self, asset_to_buy: AssetId, amount_to_buy: Balance, asset_to_pay: AssetId) -> Result<u64> {
+		let api = self.client.runtime_api();
+		let best = self.client.info().best_hash;
+		let at = BlockId::hash(best);
+
+		let result = api
+			.buy_price(&at, asset_to_buy, amount_to_buy, asset_to_pay)
+			.map_err(|e| RpcError {
+				code: ErrorCode::ServerError(Error::Runtime.into()),
+				message: "Unable to query buy price.".into(),
+				data: Some(format!("{:?}", e).into()),
+			})?;
+		match result {
+			CennzxSpotResult::Success(price) => {
+				TryInto::<u64>::try_into(price.saturated_into::<u128>()).map_err(|e| RpcError {
+					code: ErrorCode::ServerError(Error::PriceOverflow.into()),
+					message: "Price too large.".into(),
+					data: Some(format!("{:?}", e).into()),
+				})
+			}
+			CennzxSpotResult::Error => Err(RpcError {
```

Comment:
The error part of "result" is already mapped to the closure that's defined on line 97. If error and success are properly used, this arm shouldn't be reached.